lab_2/main.py

```diff
@@ -89,7 +180,30 @@ def predict_language_knn(unknown_text_vector: list, known_text_vectors: list,
     :param k: the number of neighbors to choose label from
     :param metric: specific metric to use while calculating distance
     """
-    pass
+    if not (
+        isinstance(unknown_text_vector, list)
+        and all(isinstance(n, (int, float)) for n in unknown_text_vector)
+        and isinstance(known_text_vectors, list)
+        and all(isinstance(m, list) for m in known_text_vectors)
+        and isinstance(language_labels, list)
+        and all(isinstance(s, str) for s in language_labels)
+        and len(known_text_vectors) == len(language_labels)
+        ):
+        return None
+    if metric == 'euclid':
+        calc_dist = calculate_distance
+    elif metric == 'manhattan':
+        calc_dist = calculate_distance_manhattan
+    distances = [calc_dist(unknown_text_vector, knw_vector) for knw_vector in known_text_vectors]
+    sorted_distances = (sorted(distances))[:k]
+    langs = [language_labels[distances.index(dist)] for dist in sorted_distances]
+    langs = sorted(zip(langs, sorted_distances), key=lambda x: x[1])
```

Comment:
what happens here?

Reply:
197. The list of distances is created.
For each known_vector in known_text_vectors the function calc_dist (calculate_distance or calculate_distance_manhattan) is called. calc_dist returns a float number which is put in the list 'distances'. 
e. g. distances = [0.57297, 0.849, 0.83684, 0.60025, 0.67904, 0.74431]

198. The list of k closest distances is created.
All values in the list of distances are sorted in ascending order and the k closest values are chosen. 
e.g. sorted_distances = [0.57297, 0.60025, 0.67904]

199. The list of language labels is created.
For each value of distance in sorted_distances we find its index in the initial list of values (callled 'distances'). We know that 
1) the lengths of lists 'distances' and 'language_labels' are equal 
2) and the value of distance in 'distances' corresponds to language in 'language_labels'
That is why it's possible to find the index of distance in distances and then find an appropriate language by this index.
e. g. langs = ['en', 'de', 'de'] 

200. Sorted in a certain way list of tuples is created.
The function zip () creates an iterator object from which a two-element tuple is retrieved. The first element is language label (from 'langs' list), the second element is distance (from 'sorted_distances' list). Then the tuples in the list are sorted by distance in ascending order.  
e. g. langs = [('en', 0.57297), ('de', 0.60025), ('de', 0.67904)]